site2/docs/cookbooks-message-dispatch-throttling.md

```diff
@@ -0,0 +1,105 @@
+but the sum of all the number of messages in batches may already exceed the flow threshold. 
+- Cause-2: The logic of the dispatch throttling is: `1.get remaining amount` -> `2.load data` -> `3.deduct amount`, If
+there are two process `a & b` are executed in parallel, it is possible to execute in this order: 
+`process-a: 1.get remaining amount` -> `process-a: 2.load data` -> `process-b: 1.get remaining amount` ->
+`process-b: 2.load data` -> `process-a: 3.deduct amount` -> `process-b: 3.deduct amount`, both `process-a` and
+`process-b` dispatch enough messages, and the total number exceeds the threshold.
+
+#### Estimate the amount of data to be read from the Storage
+Messages is stored in blocks. If batch-send is not enabled, each message is packaged into an independent data block.And
+if batch-send is enabled, a batch of messages are packaged into one data block, then we can not determine the true count
+of messages in one data block; We also cannot determine the true size of this data block until we have actually read the
+it from Storage.So in order to satisfy the constraints `dispatchThrottlingRateInMsg` and `dispatchThrottlingRateInByte`,
+requires a mechanism to estimate how much need to read a data block.
+
+When delivering a message to the client, the Broker looks at the size and number of messages per data block, calculates
+an average and caches it in memory, using this average to estimate how many data blocks to be read. What if there is no
+estimated value in memory when the first reading of a new topic? The current workaround is that just delivers one data
+block to clients if it is the first delivery of a new topic. Of course, the broker does not store all the information
+of data blocks for a long time to process the average accurately, which would waste memory. Instead, a compromise
+algorithm 'avg = (history-avg * 0.9 + new-avg * 0.1)' is used.
+
+> If enabled the feature `dispatchThrottlingOnBatchMessageEnabled`, will support rate-limiting dispatching on the batch
+> messages rather than individual messages within batch messages. Since one batch is one data block, this is sufficient
+> to constrain the concurrency of read requests on storage. This makes the count of the number of messages inaccurate,
+> but also maximizes pulsar's throughput while keeping storage read requests stable.
+
+#### Automatically close message dispatch throttling when there is no backlog
+After we turn on `dispatchThrottlingOnNonBacklogConsumerEnabled`(default is enabled), if all consumers in one
+subscription have no backlog(it is clear that almost all read requests can hit the cache), then message dispatch
+throttling is turned off automatically(it means that even if we set `dispatchThrottlingRateInMsg` and
+`dispatchThrottlingRateInByte`, throttling won't work because there is no backlog)，and if any consumer has backlog,
+it will be turned on automatically. If we need only to prevent the excessive read requests results in Storage can not
```

Comment:
> If we need only to prevent the excessive read requests results in Storage can not work well, this feature is very useful.

Can you clarify this part?

Reply:
I changed the instructions to this:

```
if all consumers in one subscription have no backlog(it is clear that almost all read requests can hit the cache, so Pulsar will not send read request to the Storage any more.)
```